### DISCUSSION TITLE
Native vs Cross Platform

### DISCUSSION THREAD

**Discussion by MohabTarek** [MobileNativeFoundation/discussions#43] [General]
I'm curious to hear from different teams how they choose between native and cross platform approaches, and whether this decision was based on cost reduction, business alignment or any other factors.
Do we still have those performance gaps between native and cross or not?

**Comment by barriosnahuel** [comment]
This [research](https://cdn.nubank.com.br/mobile/taskforce/nubank-mobile-architecture-task-force-mission-report.pdf) by Nubank is quite interesting. It shows many aspects that should be taken into account before going any way.

**Reply by Shadester** [reply]
Just a note that it only compares between different cross platform approaches, and not between cross platform and native.

**Comment by sregg** [comment]
We started integrating React-Native [3 years ago](https://medium.com/wealthsimple/engineering-at-wealthsimple-integrating-react-native-into-our-existing-native-apps-9b0c31627a75) at [Wealthsimple](https://wealthsimple.com/).
Here are some of our findings 3 years later:

## Benefits

### Cross-platform
This is obviously the main benefit of React-Native. We share 99% of the code we write between Android and iOS.
That means you write features once and fix bugs once.
That also means you're not shipping your Android app months later (yes, I'm looking at you Clubhouse).

### Developer Experience
_"Come for the cross-platform, stay for the dev experience"_

5 minute builds in a native project are not uncommon. With React-Native, you get instant and stateful refreshes.
You also get a great way to build UIs with React. Think Swift-UI and Jetpack Compose but with 10 years of maturity.
Its declarative and composable nature makes building a design system trivial.
Also, with React comes an amazing ecosystem of tools that takes your DX to the next level:
- Storybook (again, like Swift-UI and Jetpack Compose previews)
- Testing frameworks (jest, testing-library, etc...)
- Dependency management (npm is much easier to understand than gradle or cocoapods)
- etc...

Finally, if you don't like Javascript because it's not typed, you can always use Typescript (very similar to Swift and Kotlin).

### Learn once, write everywhere
Because React-Native is built on React, your web developers can also write code for your mobile app.
This makes a vertical organization of your engineering team much easier to implement.

### Battle-tested
Facebook is not the only company using React-Native. Here's a non-exhaustive list of other big companies using it:
[Coinbase](https://blog.coinbase.com/optimizing-react-native-7e7bf7ac3a34), [Shopify](https://engineering.shopify.com/blogs/engineering/react-native-future-mobile-shopify), [Discord](https://blog.discord.com/how-discord-achieves-native-ios-performance-with-react-native-390c84dcd502), [Microsoft](https://appfigures.com/resources/insights/microsoft-goes-all-in-on-react-native), Tesla, and [many more](https://reactnative.dev/showcase).

## Things to consider

### You still need native experience
90% of the time, you'll be able to stay in the JS land and not think about the native world across the bridge.
When you do (e.g. push notifications, app launch, etc...), only developers with mobile knowledge will be able to help here.

### The JS ecosystem
The JS ecosystem, with its many packages and tools, evolves so rapidly that things can break sometimes.
Mobile devs coming from Android and iOS native development will have to get familiar with a new way of doing things: new language, new IDE, new dependency management, linting, and building tools, etc...

### Brownfield vs greenfield
We started by integrating RN incrementally into our first app.
This was much harder to execute than to start a new project with RN from scratch (which we did for the 2 new apps).

**Reply by KenLaam** [reply]
>  When you do (e.g. push notifications, app launch, etc...), only developers with mobile knowledge will be able to help here.

totally agreed, I'm native developer for both iOS and Android. Now, I'm developing ReactNative app. Yes, ReactNative app, my main tasks are related to Push Notification, multi background thread, deep link, and hardware things such as Camera, Microphone

TBH, you should have backup about pure native

**Comment by theamorn** [comment]
Want to share knowledge from my Flutter Experience with a high-security mobile banking app for 1.5 years. Copy template from above :D
Some background about me. iOS developer around 11 years, Android 1 year and never work with React Native.

**Cross-platform**
Flutter shares around 95% of the code between iOS/Android, the rest is implemented on Native because Flutter lib does not fully support our requirements, mostly security things. We add a lot of Flutter Bridge to Native for calling 3rd party native SDK which Flutter doesn't have

**Developer Experience**
It takes a lot of time to build an app for iOS, From a cold start, it took like 3-4 minutes in my project while iOS with Xcode take around 2 minutes
In Android, it's a lot faster and mostly I test and build the app in Android instead due to the faster build time
Android took around 40-50 seconds for a cold start, while Android Studio took about the same time.
Hot reload is a game-changer for a mobile developer, Hot reload took 2-3 seconds to change the UI while iOS/Android is not mature enough to be useful for now. I try with SwiftUI, and Instant run already.
Since business logic is the same so no headache about wrong implementation on the specific platform. Tester has more time to test business logic and correctness of the features than testing if iOS/Android implements it right? It's the same code base, If it's right it will both right, if it's wrong, both will be wrong.

**CI/CD**
it's basically iOS/Android, since all Flutter app will be built into native so we can use CI/CD for iOS/Android pipeline without any problem

**Testing**
Unit testing is fine, nothing special like other language. Work as intended. but the problem come with UI Testing.
Since Xcode/Android Studio has tools for Record mode, so even non-dev can still working on UI Test without much knowledge in development, while Flutter UI Test is totally different. Everything needs to be in the code, no Record Mode :(

**Learn once, write everywhere**
Flutter just release Flutter 2.0 with production-grade for web development, Don't try it yet, but it should be a game-changer as well that writes once, deploy everywhere in the really same code base. It's the same concept of Adobe Air that I tried once but different performance

**Battle-tested**
Google back it up and Google Pay use it to showcase, right now it has around 150,000 apps so far still growing with a faster growth rate than other Cross-platform
https://developers.googleblog.com/2020/09/google-pay-picks-flutter-to-drive-its.html

**Things to consider**
Yes, From my experience, you saved over 43% of development cost, not 50% like in theory. 
because you have to spend time on technical debt like fix some weird bugs that Flutter has, hack around to make it works. But mostly it should be cosmetic bugs, not fatal ones so it's still save your cost.
43% save is just for 2 Platform, iOS/Android. What if your business needs 4-5 platforms? it will save even more.

It's the same as React Native, You still need a native mobile experience.
In my opinion, I prefer native mobile dev to use Flutter instead of web developer because of the app life cycle, certificates, permission handler. They will learn a lot faster. That just my suggestion, in the end, the web developer can learn it as well but just take a longer time

**Performance**
Basically, it's compiled dart to binary native. so It's native. But still, have some bugs to make animation laggy but 95% of the app should be fine. No need to optimize performance to look like Native that much, but some part if you feel that it's not right, you can optimize it

**Verdict**
if your app is a ListView-to-Details app and 90 - 95 % of the app in the market are these types of app. You can use Flutter without any problem.
The app that can't use is high CPU intensive calculation such as VDO editing, Photo editing app, Gaming.
You can develop a VDO Editing app but You will lose a chance to make it competitive
For example
VDO Editing render 4K for in Native - took 30 seconds
VDO Editing render 4K for in Native - took 50 seconds
Your Flutter app won't have a chance to win Native at all.

Flutter is Native by theory but not in the real world.

**Reply by KenLaam** [reply]
thanks for sharing

**Reply by XabierGoros** [reply]
Thanks for sharing @theamorn. There's a small typo in: "VDO Editing render 4K for in **Flutter** - took 50 seconds".

**Comment by ryanggrey** [comment]
This is a great question and the answers provide very good information. I feel like there's a lot of information out there in blogs and discussion boards with various degrees of fragmentation, completeness and how current the information is. I also think that this is a question that is repeatedly asked as organisations and teams look to start new products or optimise their current multi-platform mobile products. This means that there's likely a lot of duplicated work done by organisations in answering this question.

Does anyone know if there is a framework out there for helping organisations and teams through the technology selection process for a cross-platform product? And if not is it something you think organisations would find useful?

I realise how difficult it would be to encompass such a broad and subjective topic in a useful manner, but a framework could be focused on helping an organisation with asking the right questions of their product/project rather than providing opinionated recommendations. 

For example something similar to [the way Nubank made their assessment](https://github.com/MobileNativeFoundation/discussions/discussions/43#discussioncomment-434237) but in a generic form that allows an organisation to go through an assessment in a structured manner. Another example of what this might look like at a very high-level are the [checklists](https://github.com/OWASP/owasp-mstg/tree/master/Checklists) and resources that the [OWASP mobile security testing guide](https://owasp.org/www-project-mobile-security-testing-guide/) provides to help users perform security assessments of their apps (I realise security is a different topic entirely and only point to it as an example of a framework that helps an organisation through a relatively in-depth self-review/assessment of a topic).

Any thoughts on the plausibility of this or how useful it could be for organisations?

**Comment by rudro** [comment]
Adding for completeness:
https://medium.com/airbnb-engineering/react-native-at-airbnb-f95aa460be1c (from mid-2018)
and 
https://shopify.engineering/react-native-future-mobile-shopify (roughly a year ago)

**Comment by piannaf** [comment]
I've been talking with and researching exactly this for large and small engineering orgs because of the work I do with @touchlab.

Here's a selection of some reasons I've heard for choosing Kotlin Multiplatform as a native code sharing solution vs a cross-platform solution
* Must have strong offline capabilities
* Heavy reliance on device sensors
* Been burned by choosing cordova/react-native/xamarin in the past
* Don't want to reskill the team
* Want to focus on sharing business logic and don't want to retool UIs
* Maximizing the the value of code sharing is better than maximizing the amount of code sharing
* The risk of making the wrong choice (backing out of KMP can keep Android app, backing out of others means starting over)
* See the value of native UI for optimal customer experience
* Retain great devs who see the value in learning and using Swift, SwiftUI, Kotlin, and Jetpack Compose
* Apple's mobile investments and Google's Android investments are a more stable bet than Google's Flutter and other 3rd party solutions
* Don't want to introduce a 3rd platform (still need native devs with other solutions)
* Can use innovations from Apple and Google as soon as appropriate rather than waiting on 3rd party or community
* Current library choices don't have equivalents in other solutions, are less maintained, featureful, or would require taking on the work of making them work

Here are some more articles ([airbnb](https://github.com/MobileNativeFoundation/discussions/discussions/43#discussioncomment-456753), [shopify](https://github.com/MobileNativeFoundation/discussions/discussions/43#discussioncomment-456753), and [nubank](https://github.com/MobileNativeFoundation/discussions/discussions/43#discussioncomment-434237) already in previous comments)
* [VMWare chooses KMP](https://medium.com/vmware-end-user-computing/adopting-a-cross-platform-strategy-for-mobile-apps-59495ffa23b0)
* [Square chooses KMP](https://developer.squareup.com/blog/developing-on-ios-and-android/)
* [Quizlet chooses KMP](https://medium.com/tech-quizlet/shared-code-at-quizlet-kotlin-multiplatform-2ee1b57646c)
* [Wantedly moving from React Native to KMP](https://medium.com/wantedly-engineering/moving-from-react-native-to-kotlin-multiplatform-292c7569692)
* [Udacity moving from React Native to Native](https://engineering.udacity.com/react-native-a-retrospective-from-the-mobile-engineering-team-at-udacity-89975d6a8102)
* [infinut Math chooses KMP over Unity](https://medium.com/@anaredmond/unity-vs-kotlin-native-64cd75386357)
* [Netflix chooses KMP](https://netflixtechblog.com/netflix-android-and-ios-studio-apps-kotlin-multiplatform-d6d4d8d25d23)
* [SuperAwesome chooses KMP](https://www.superawesome.com/blog/how-we-developed-our-new-video-platform-rukkaz-as-a-cross-platform-mobile-app/)
* [Slack moves away from C++](https://slack.engineering/client-consistency-at-slack-beyond-libslack/)
* [Dropbox moves away from C++](https://dropbox.tech/mobile/the-not-so-hidden-cost-of-sharing-code-between-ios-and-android)

The Slack and Dropbox ones are really interesting one because a lot of people consider C++ native. But that only takes into account the machine it's running on, not the developer or user experience. The fact that you need a bridge via NDK for C++ on Android means C++ is not native for Android application development. As I described in one of my talks [The future of cross-platform is Native](https://youtu.be/sA_JIqqj9js)
![image](https://user-images.githubusercontent.com/70764/110997339-40548980-834b-11eb-8822-421f623f6b0b.png)

I also talk about some of the decision making around sticking with Native, choosing code sharing, or x-platform in another talk [Mobile Cross-Platform Development: Picking the Right Solution (Kotlin, React Native, Flutter)](https://youtu.be/6RquJJM1jaE)

**Reply by sregg** [reply]
Can you write UIs with KMP?
Also, how do iOS devs react when they hear they have to learn Kotlin?

**Reply by Alex009** [reply]
hi @sregg !
In common case UI implemented on android and ios separately, with standard toolkit. but also we at IceRock already try to implement UI in common code. As result now we have https://github.com/icerockdev/moko-widgets (but for now project is frozen). also jetpack compose migration on iOS already have proof of concept - https://github.com/cl3m/multiplatform-compose

but in production apps i suggest to use platform native UI - for now common UI at KMP have many limitations.

our iOS devs react on Kotlin without entusiasm at start, and iOS devs not write Kotlin code - only use framework compiled from kotlin code. But now (after 2.5 years with KMP) iOS devs implement common code too.

**Reply by kpgalligan** [reply]
> Can you write UIs with KMP?

Eventually there will be more UI options, although most of the use cases where you'd be looking at KMP aren't "all-in-one" type of solutions vs code sharing of modules within a larger app, or for smaller apps, sharing the majority of the logic below the UI layer.

There are attempts to create a unified UI layer, similar to the kinds of things that happen on React Native. The actual UI is rendered from native components with a common DSL on top of it (see icerock post above, which does this AFAIK).

There are people poking around Compose, and I think that'll eventually be an option, but it'll be a while (I've been poking around, and I can tell you first hand, it'll be a while).

I would also expect more purpose-built UI layers (forms, "Design system specific" implementations, etc).

> Also, how do iOS devs react when they hear they have to learn Kotlin?

Well, this thread started with the 100% native vs "cross-platform", which is not a reasonable comparison for KMP. To stick with the UI discussion, I could reframe this question with:

> How do iOS devs react when they find out they can write UIs in SwiftUI (or UIKit) but get to avoid using CoreData for databases?

KMP is "native code sharing", not a "cross platform toolkit". We work with a number of larger teams in which most of the iOS devs won't ever need to edit Kotlin, but can ease into it if they want (or their manager tells them too I guess :).

We (touchlab) also publish an [Xcode Kotlin](https://github.com/touchlab/xcode-kotlin) plugin that will let you browse and debug Kotlin running in iOS. If you're going to edit Kotlin, you'll really need Android Studio or Intellij.

On Kotlin itself, Kotlin is syntactically very similar to Swift. They are, however, different in a number of details, and missing some of those features in Kotlin can be an issue for Swift devs. There needs to be an expectation of some compromise or none of these options will work. "how do iOS devs react when they hear they have to learn Kotlin?", I'd say pretty good compared to how they might react when they hear they need to code in JS or Dart.

Summary: KMP is an option, and it's easy to use within an existing native app, or reduce your "platform risk" with new apps.  You need to use it in the right context, though. And, for sure, it's possible not everybody on the team will love it.
